openshift-kube-apiserver/openshiftkubeapiserver/patch_handlerchain.go

```diff
@@ -95,3 +105,104 @@ func translateLegacyScopeImpersonation(handler http.Handler) http.Handler {
 		handler.ServeHTTP(w, req)
 	})
 }
+
+// withWatchRateLimit rate limits the establishment of watches to keep kube-apiservers from crashing.
+// Rate limiting watches effectively creates an upper bound on secret and configmap mounts of 10*QPS because of a
+// ten minute watch timeout and kubelets use watches to get the content for the mount.
+// We will break the rate limiting into three timespans
+// 1. first ten minutes: this is the most restrictive 10000 total mounted secrets and configmaps, 1000 per minute.
+//    We're trying to break up the slug of kubelet traffic and
+//    we want to be sure that operators can make progress during this time if we need to recover a cluster in
+//    a bad state.
+// 2. second ten minutes: this is less restrictive  20000 total mounted secrets and configmaps, 2000 per minute.
+//    This lets us start to ramp up during a relative steady state.
+// 3. no limit.  We have this to handle cases of large clusters with more than 20000 mounted secrets and configmaps.
+//    I honestly don't know how common this is, but I don't want to break on it.
+// Recall that we observed more than the the 30,000 per minute observed during some disruptive events on a cluster.
+// In addition, we special case watches in the platform operator namespaces, cluster scope, and kube-system.
+// We have not observed large numbers of these and they are required in order to make progress when trying to correct
+// some kinds of cluster failures.
+func withWatchRateLimit(handler http.Handler) http.Handler {
+	earlyRateLimiter := flowcontrol.NewTokenBucketRateLimiter(16.6, 100)
+	middleRateLimiter := flowcontrol.NewTokenBucketRateLimiter(33.3, 100)
+
+	// after 20 minutes, we eliminate the rate limiter.  This lets us tolerate very large clusters, while having a ramp
+	// up period during which we can make forward progress
+	startTime := time.Now()
+	earlyEndTime := startTime.Add(10 * time.Minute)
+	middleEndTime := startTime.Add(20 * time.Minute)
+
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		var effectiveRateLimiter flowcontrol.RateLimiter
+		now := time.Now()
+		switch {
+		case now.After(middleEndTime):
+			// we are past our rate limiting time
+			handler.ServeHTTP(w, req)
+			return
+		case now.After(earlyEndTime):
+			effectiveRateLimiter = middleRateLimiter
+		default:
+			effectiveRateLimiter = earlyRateLimiter
+		}
+
+		ctx := req.Context()
+
+		attributes, err := filters.GetAuthorizerAttributes(ctx)
```

Comment:
what about watches to the aggregated APIs? Should they also be rate-limited?

Reply:
And by whom?

Reply:
> what about watches to the aggregated APIs? Should they also be rate-limited?

They are.



> And by whom?

By this.

The filter chain runs before all apiservers

Reply:
I was asking if they should be rate-limited by this PR

Reply:
or should we add rate-limiting to individual apis?

Reply:
> I was asking if they should be rate-limited by this PR

I think this enforcement is ok.  It's not perfect, but I think it's reasonable.